### PR TITLE
ENH: slurm launcher

### PR DESCRIPTION
**Pull request by jbwexler**
Sorry for the delay, I've finally been able to work on this. This fix turns out to be pretty simple and seems to work well so far but I haven't done any kind of extensive testing yet. Just thought I would let you guys take a look. Let me know if you have questions. I think there's probably a more elegant way of creating the SlurmLauncherSubmitter class than how did it, which was simply to copy and paste the SlurmSubmitter class and make a few minor edits.

As I write this it occurs to me that we should also add some of the if statements in slurm_launcher.template to slurm.template in order to fix https://github.com/ReproNim/reproman/issues/549 

Some example output:

> (main) jbwexler@login2>reproman run --follow -r local --sub slurm_launcher --orc datalad-no-remote --batch-parameter sub=1,2,3,4 --input sub-0{p[sub]} --jp num_processes=4 --jp num_nodes=4 --jp walltime=00:02:00 --jp queue=normal sh ../test.sh
> 2021-02-25 13:57:49,928 [INFO   ] Making inputs available
> 2021-02-25 13:57:50,153 [INFO   ] Submitting 20210225-135748-648d
> 2021-02-25 13:57:50,158 [INFO   ] No root directory supplied for local; using '/home1/03201/jbwexler/.reproman/run-root'
> 2021-02-25 13:57:50,314 [INFO   ] Submitting /scratch/03201/jbwexler/reproman_test/ds000001/.reproman/jobs/local/20210225-135748-648d/submit
> 2021-02-25 13:57:50,561 [INFO   ] Job 20210225-135748-648d submitted as slurm_launcher job 3286023
> 2021-02-25 13:57:50,571 [INFO   ] Registered job 20210225-135748-648d
> 2021-02-25 13:57:50,596 [INFO   ] Waiting on job 3286023: PENDING. Next heartbeat in 5 seconds
> 2021-02-25 13:57:55,856 [INFO   ] Waiting on job 3286023: PENDING. Next heartbeat in 5 seconds
> 2021-02-25 13:58:01,730 [INFO   ] Waiting on job 3286023: PENDING. Next heartbeat in 12 seconds
> 2021-02-25 13:58:14,224 [INFO   ] Waiting on job 3286023: PENDING. Next heartbeat in 26 seconds
> 2021-02-25 13:58:45,751 [INFO   ] Waiting on job 3286023: PENDING. Next heartbeat in 60 seconds
> 2021-02-25 13:59:49,777 [INFO   ] Waiting on job 3286023: PENDING. Next heartbeat in 131 seconds
> 2021-02-25 14:02:06,773 [INFO   ] Waiting on job 3286023: PENDING. Next heartbeat in 281 seconds
> 2021-02-25 14:06:53,677 [INFO   ] Waiting on job 3286023: PENDING. Next heartbeat in 597 seconds
> 2021-02-25 14:16:51,819 [INFO   ] Waiting on job 3286023: PENDING. Next heartbeat in 1255 seconds
> 2021-02-25 14:25:43,272 [INFO   ] Final state of job 3286023: COMPLETING
> 2021-02-25 14:25:43,339 [INFO   ] Unregistered job 20210225-135748-648d

### CODE REVIEW

**Comment by yarikoptic** [comment]
Great, thank you @jbwexler !  

I wonder if it is easyiesh to install/"provision" the submitter on top of an existing slurm instance?  
If so -- it would be great to get some "unit" test to ensure that it is all working.  We have https://github.com/ReproNim/reproman/blob/master/tools/ci/setup-slurm-container.sh and some testing of interactions with slurm in https://github.com/ReproNim/reproman/blob/master/reproman/support/jobs/tests/test_orchestrators.py

@kyleam could you please review the PR?

Re

> As I write this it occurs to me that we should also add some of the if statements in slurm_launcher.template to slurm.template in order to fix #549

would be much appreciated  -- might better be in a separate PR to not make it contingent on this one.

File: reproman/support/jobs/job_templates/submission/slurm_launcher.template [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+{% endif %}
+{% if queue is defined %}
+#SBATCH -p {{ queue }}
+{% endif %}
+
+module load launcher
+export LAUNCHER_JOB_FILE="{{ shlex_quote(_meta_directory) }}/${SLURM_JOB_ID}"
+
+for task_id in {0..{{ _num_subjobs - 1}}}; do
+    echo "{{ shlex_quote(_meta_directory) }}/runscript $task_id" >> $LAUNCHER_JOB_FILE
```
Comment (kyleam) [inline reproman/support/jobs/job_templates/submission/slurm_launcher.template:33]: `"$LAUNCHER_JOB_FILE"` just to be safe, and the same for `$LAUNCHER_DIR` below.

File: reproman/support/jobs/job_templates/submission/slurm_launcher.template [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+#SBATCH -t {{ walltime }}
+{% endif %}
+{% if queue is defined %}
+#SBATCH -p {{ queue }}
+{% endif %}
+
+module load launcher
+export LAUNCHER_JOB_FILE="{{ shlex_quote(_meta_directory) }}/${SLURM_JOB_ID}"
+
+for task_id in {0..{{ _num_subjobs - 1}}}; do
```
Comment (kyleam) [inline reproman/support/jobs/job_templates/submission/slurm_launcher.template:32]: This file is currently `bin/sh`, so I'd prefer `$(seq 1 x)` to `{0..x}`.

**Review by kyleam — COMMENTED** [review]
Thanks @jbwexler.

> I think there's probably a more elegant way of creating the SlurmLauncherSubmitter class than how did it, which was simply to copy and paste the SlurmSubmitter class and make a few minor edits.

Hmm, it looks like only `.name` is changing between the classes:

```diff
diff --git a/orig b/launcher
index ccd263e12..cc815c64e 100644
--- a/orig
+++ b/launcher
@@ -1,19 +1,18 @@
-class SlurmSubmitter(Submitter):
-    """Submit a Slurm job.
+class SlurmLauncherSubmitter(Submitter):
+    """Submit a Slurm Launcher job.
     """
-    name = "slurm"
+    name = "slurm_launcher"
 
     def __init__(self, session):
-        super(SlurmSubmitter, self).__init__(session)
+        super(SlurmLauncherSubmitter, self).__init__(session)
 
     @property
     @borrowdoc(Submitter)
     def submit_command(self):
         return ["sbatch"]
-
     @borrowdoc(Submitter)
     def submit(self, script, submit_command=None):
-        out = super(SlurmSubmitter, self).submit(script, submit_command)
+        out = super(SlurmLauncherSubmitter, self).submit(script, submit_command)
         # Output example (v19.05): Submitted batch job 5
         job_id = out.strip().split()[-1]
         self.submission_id = job_id
```

So that duplicated class is just serving to link up the right submission template file:

https://github.com/ReproNim/reproman/blob/6c49cfeca9ebb6217a1f2de8d1152a9b61af5768/reproman/support/jobs/orchestrators.py#L250-L253

One option then would be to allow the template name to be specified and stick with `SlurmSubmitter`.  Another (my current preference) would be to stick with `SlurmSubmitter` *and* `slurm.template`, adding a job parameter that means "use the launcher".  Each relevant submission template could then conditionally `include` the relevant launcher template.  (`reproman/support/jobs/job_templates/runscript/` has some `{% include ... %}` examples.)  At the very least, that'd mean absorbing this PR's `slurm_launcher.template` into `slurm.template`, but IIUC `launcher` isn't tied to SLURM, so perhaps it'd make sense to use it in some other submission templates as well.

> As I write this it occurs to me that we should also add some of the if statements in slurm_launcher.template to slurm.template in order to fix #549

I think adding the job parameter like you did is a good solution, and the parameter probably could be used in some of the other submission templates as well.

**Comment by jbwexler** [comment]
Thanks for the feedback! @kyleam I like your solution and am working on it now. 

One question: I need to add --ntasks to slurm.template, which can take a job parameter called 'num_tasks'. My slight concern is that [some documentation](https://documentation.sigma2.no/jobs/job_scripts/slurm_parameter.html) describes --ntasks as the number of processes, which is confusing considering reproman has a different job parameter called num_processes, which is assigned to --cpus-per-task. For what it's worth, the above documentation describes --cpus-per-task as the number of threads. My understanding of the differences between procs/threads/tasks is limited so I thought I would get your opinion on this.

**Comment by kyleam** [comment]
That's a good question, and I think in general how to handle these parameters across the different submitters is an unresolved issue:

https://github.com/ReproNim/reproman/blob/6c49cfeca9ebb6217a1f2de8d1152a9b61af5768/reproman/support/jobs/job_templates/submission/slurm.template#L5-L11

I think it was a bad choice to map `num_processes` to `--cpus-per-task`.  The Slurm template came after PBS and Condor.  For PBS it maps to `ppn`, so I guess `--ntasks-per-node` (without `--ntasks`) would be the closest match.  For Condor it is `request_cpus` and maybe `--ntasks` would be the best fit.  My suggestion for now would be to drop the use of `--cpus-per-tasks` and go with `--ntasks` or `--ntasks-per-node`.

**Comment by kyleam** [comment]
Fwiw I've added launcher to the container that we test the current slurm submitter against.  Based on quick testing, it looks like that may be enough to work with your new template.  I'll try to wire that up after you push the next iteration.

https://github.com/ReproNim/reproman-slurm/commit/1cd5e8859e31d72fec4c1b3542d7659e018c7b2e

**Comment by jbwexler** [comment]
I ended up creating a new branch here: https://github.com/ReproNim/reproman/pull/577

**Pull request closed without merging**